src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,270 @@
+                // actually searching for an alias, they can always get that information out in the end
+                // by checking the ReferenceLocations that are returned.
+                var searchSymbol = symbol;
+
+                if (searchSymbol is IAliasSymbol aliasSymbol)
+                    searchSymbol = aliasSymbol.Target;
+
+                searchSymbol = searchSymbol.GetOriginalUnreducedDefinition();
+
+                // If they're searching for a delegate constructor, then just search for the delegate
+                // itself.  They're practically interchangeable for consumers.
+                if (searchSymbol.IsConstructor() && searchSymbol.ContainingType.TypeKind == TypeKind.Delegate)
+                    searchSymbol = symbol.ContainingType;
+
+                Contract.ThrowIfNull(searchSymbol);
+
+                // Attempt to map this symbol back to a source symbol if possible as we always prefer the original
+                // source definition as the 'truth' of a symbol versus seeing it projected into dependent cross language
+                // projects as a metadata symbol.  If there is no source symbol, then continue to just use the metadata
+                // symbol as the one to be looking for.
+                var sourceSymbol = await SymbolFinder.FindSourceDefinitionAsync(searchSymbol, solution, cancellationToken).ConfigureAwait(false);
+                return sourceSymbol ?? searchSymbol;
+            }
+
+            /// <summary>
+            /// Determines the initial set of symbols that we should actually be finding references for given a request
+            /// to find refs to <paramref name="symbol"/>.  This will include any symbols that a specific <see
+            /// cref="IReferenceFinder"/> cascades to, as well as all the linked symbols to those across any
+            /// multi-targetting/shared-project documents.  This will not include symbols up or down the inheritance
+            /// hierarchy.
+            /// </summary>
+            private static async Task<HashSet<ISymbol>> DetermineInitialSearchSymbolsAsync(
+                FindReferencesSearchEngine engine, ISymbol symbol, CancellationToken cancellationToken)
+            {
+                var result = new HashSet<ISymbol>();
+                var workQueue = new Stack<ISymbol>();
+
+                // Start with the initial symbol we're searching for.
+                workQueue.Push(symbol);
+
+                // As long as there's work in the queue, keep going.
+                while (workQueue.Count > 0)
+                {
+                    var currentSymbol = workQueue.Pop();
+                    await AddCascadedAndLinkedSymbolsToAsync(engine, currentSymbol, result, workQueue, cancellationToken).ConfigureAwait(false);
+                }
+
+                return result;
+            }
+
+            private static async Task<HashSet<ISymbol>> DetermineInitialUpSymbolsAsync(
```

Comment:
Are there cases where we add up symbols that aren't part of the initial? Like through linked symbols?

Reply:
yes.  a simply example is if you start on an override.  The 'initial up' symbols will contain the member that this override overrode (And everything higher).